tests/src/common/test_case/app.rs

```diff
@@ -0,0 +1,141 @@
+        self.app.store_code(code)
+    }
+
+    pub fn time_shift(&mut self, duration: Duration) {
+        self.app.time_shift(duration)
+    }
+
+    pub fn update_block<F>(&mut self, f: F)
+    where
+        F: Fn(&mut BlockInfo),
+    {
+        self.app.update_block(f)
+    }
+
+    #[must_use]
+    pub fn block_info(&self) -> BlockInfo {
+        self.app.block_info()
+    }
+
+    pub fn send_tokens(
+        &mut self,
+        sender: Addr,
+        recipient: Addr,
+        amount: &[CwCoin],
+    ) -> anyhow::Result<()> {
+        self.app
+            .send_tokens(sender, recipient, amount)
+            .map(|_: AppResponse| ())
+    }
+
+    pub fn instantiate<'r, T, U>(
+        &'r mut self,
+        code_id: u64,
+        sender: Addr,
+        init_msg: &T,
+        send_funds: &[CwCoin],
+        label: U,
+        admin: Option<String>,
+    ) -> anyhow::Result<ResponseWithInterChainMsgs<'r, Addr>>
+    where
+        T: Debug + Serialize,
+        U: Into<String>,
+    {
+        self.with_mock_app(|app: &mut MockApp| {
+            app.instantiate_contract(code_id, sender, init_msg, send_funds, label, admin)
+        })
+    }
+
+    pub fn execute<'r, T>(
+        &'r mut self,
+        sender: Addr,
+        contract_addr: Addr,
+        msg: &T,
+        send_funds: &[CwCoin],
+    ) -> anyhow::Result<ResponseWithInterChainMsgs<'r, AppResponse>>
+    where
+        T: Debug + Serialize,
+    {
+        self.with_mock_app(|app: &mut MockApp| {
+            app.execute_contract(sender, contract_addr, msg, send_funds)
+        })
+    }
+
+    pub fn execute_raw<T>(
+        &mut self,
+        sender: Addr,
+        msg: T,
+    ) -> anyhow::Result<ResponseWithInterChainMsgs<'_, AppResponse>>
+    where
+        T: Into<CosmosMsg>,
+    {
+        self.with_mock_app(|app: &mut MockApp| app.execute(sender, msg.into()))
+    }
+
+    pub fn sudo<'r, T, U>(
+        &'r mut self,
+        contract_addr: T,
+        msg: &U,
+    ) -> anyhow::Result<ResponseWithInterChainMsgs<'r, AppResponse>>
+    where
+        T: Into<Addr>,
+        U: Serialize,
+    {
+        self.with_mock_app(|app: &mut MockApp| app.wasm_sudo(contract_addr, msg))
+    }
+
+    pub fn with_mock_app<F, R>(&mut self, f: F) -> anyhow::Result<ResponseWithInterChainMsgs<'_, R>>
+    where
+        F: FnOnce(&'_ mut MockApp) -> anyhow::Result<R>,
+    {
+        assert_eq!(self.message_receiver.try_recv().ok(), None);
+
+        match f(&mut self.app) {
+            Ok(result) => Ok(ResponseWithInterChainMsgs::new(
+                &mut self.message_receiver,
+                result,
+            )),
+            Err(error) => {
+                // On error no messages should be "sent out".
+                while self.message_receiver.try_iter().next().is_some() {}
+
+                Err(error)
+            }
+        }
+    }
+
+    #[must_use]
+    pub fn query(&self) -> QuerierWrapper<'_, Empty> {
+        self.app.wrap()
+    }
```

Comment:
realizing the following is a bit out-of-scope, we can simplify the query-like code in 10x test code locations. We'll avoid the dependency on cosmwasm_std::QuerierWrapper by mimicking its API here and removing this method.

Reply:
While yes, I'd prefer this to be done in a separate PR, so we can merge this and focus on the tests themselves first.